uswds/uswds
# USWDS - File Input: Allow custom status and instructions text

# Summary

**Added optional custom text data attributes to the file input component.** These attributes make it simpler to translate component text into other languages.

## New attributes

| Attribute name  | Default | 
| ---------------- | :-------------- |
| data-change-file-text |  "Change files" |
| data-change-file-text-singular | "Change file" |
| data-choose-text | "choose from folder" | 
| data-drag-text | "Drag files here or" |
| data-drag-text-singular | "Drag file here or" |
| data-error-message | "This is not a valid file type." |
| data-no-file-text | "No files selected." |
| data-no-file-text-singular | "No file selected." |
| data-selected-file-text | "files selected" |
| data-selected-file-text-singular | "file selected" |

## Breaking change

This is not a breaking change.

## Related issue

Closes #5238

## Related pull requests

[Changelog PR](https://github.com/uswds/uswds-site/pull/2300)

## Preview link

- [File input component
](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-file-input-data-drag-text/?path=/story/components-form-inputs-file-input--default)
- [File input component - custom text test](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-file-input-data-drag-text/?path=/story/components-form-inputs-file-input--test-custom-text)

## Problem statement

The file input component hard codes English strings into the JavaScript for the:
- introductory instructions
- file status
- change file instructions
- error notification

This makes it difficult for users to translate the component into other languages. 

## Solution
This PR restructures how we program and structure the copy in the component. The code now relies on full phrase strings instead of adjusting and inserting individual words. This provides the most flexibility in language and sentence structure. 

Users will need to customize 10 data attributes in order to fully translate the component text to another language. 

The JavaScript provides a minimal about of fallback where it will use the plural version of the string in place of the singular if the singular is not provided. 

> **Note**
> This PR does not yet have unit tests.

## Testing and review

1. Confirm the data attributes are intuitively named
2. Confirm the data attributes work as expected
3. Confirm that the component works without the data attributes
4. Confirm no behavioral regressions

In the [Test custom text story](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-file-input-data-drag-text/?path=/story/components-form-inputs-file-input--test-custom-text):

1. To test drag text:
    1. Set the “multiple” control to “false”. Confirm that component delivers the singular version of the initial instructions (”Arrastre *archivo* aquí o elegir de la carpeta”)
    2. Set the “multiple” control to “true” and reload. Confirm that component delivers the plural version of the initial instructions (”Arrastre *archivos* aquí o elegir de la carpeta”)
2. To test the file status text…
    1. With default values: 
        1. Set the “multiple” control to “true”, the “include_singular” control to false and reload. 
    2. With additional singular customizations:
        1. Set the “multiple” control to “true”, the “include_singular” control to true and reload
        2. Add any 1 file to the component via click or drag
            1. Confirm that the the visual presentation of the component includes the text “1 archivo seleccionado” and “Cambiar archivo”
            2. In the web inspector, confirm that the `usa-sr-only` element receives the text “1 archivo seleccionado: [File name]”
        3. Click on the component to change files. Select any 2 files.
            1. Confirm that the the visual presentation of the component includes the text “2 archivos seleccionados” and “Cambiar archivos”
            2. In the web inspector, confirm that the `usa-sr-only` element receives the text “2 archivos seleccionados: [File name], [File name]”

In the [default file input story](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-file-input-data-drag-text/?path=/story/components-form-inputs-file-input--default): 

1. To test drag text:
    1. Set the “multiple” control to “false”. Confirm that component delivers the singular version of the initial instructions (”Drag file here or choose from folder”)
    2. Set the “multiple” control to “true” and reload. Confirm that component delivers the plural version of the initial instructions (”Drag files here or choose from folder”)
2. To test the file status test:
    1. Set the “multiple” control to “true” and reload
    2. Add any 1 file to the component via click or drag
        1. Confirm that the the visual presentation of the component includes the text “1 file selected” and “Change file”
        2. In the web inspector, confirm that the `usa-sr-only` element receives the text “1 file selected: [File name]”
    3. Click on the component to change files. Select any 2 files.
        1. Confirm that the the visual presentation of the component includes the text “2 files selected” and “Change files”
        2. In the web inspector, confirm that the `usa-sr-only` element receives the text “2 files selected: [File name], [File name]”
   4. Remove selected files from the component by opening the file dialog and clicking "cancel"
       1. Confirm that the visual presentation of component returns to the initial instructions
       2. In the web inspector, confirm that the `usa-sr-only` element receives the text “No files selected.”
